Example: Cursor widget with text

Joschua-Conrad opened the pull request:
## PR Summary

Added a new example of a cursor widget, which sticks to the plotted line and displays the closest datapoint to which the cursor currently sticks.

This is the continuation of closed PR #13933, where the feature was introduced inside matplotlib's API. The example now just adds the new examples/widgets/text_cursor.py and leaves all other files of MPL unchanged.

The example is supposed to be used as cursor for interactively investigating plots and to learn about the widget API.

![text_cursor_screenshot](https://user-images.githubusercontent.com/32811308/115192212-cbfbcb80-a0ea-11eb-9640-bfd8dfc52703.png)

## PR Checklist

- [N/A] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

smartlixx commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+Textcursor
+==========
+
+Display a data cursor including a text box, which shows the plot point, which
+is close to the mouse pointer.
```
```suggestion
Display a data cursor including a text box, which shows the plot point close
to the mouse pointer.
```

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+The new cursor inherits from `~matplotlib.widgets.Cursor` and demonstrates the
+creation of new widgets and their event callbacks.
+
+.. note::
+    The figure related to this example does not show the cursor, because that
+    figure is automatically created in a build queue, where the first mouse
+    movement, which triggers the cursor creation, is missing.
```
You could send a first fake event. But not sure if it's worth  it.

Joschua-Conrad replied:
See the commit e989183725dc3e0091ea8ea74a6701ea730b32ea where I removed the following code

```python
# In the gallery picture, we cannot see the cursor.
# The cursor is displayed after the first mouse move event.
# There is no mouse move when using automatic scripts forgenerating the
# documentation. The onmove event needs to occur after the plt.show() command,
# because this one creates a fresh and clean canvas. Without the cursor or
# its text. And after the plt.show(), we cannot execute more instructions.
# The following code draws the figure (without showing it) and calls the mouse
# event manually. The resulting figure is saved to a file. But it cannot be
# shown, because this first clears the canvas and therefore removes the cursor.
from matplotlib.backend_bases import MouseEvent
location = np.array([0, 10])
location = ax.transData.transform(location)
event = MouseEvent(name='motion_notify_event', button=None, key=None,
   x=location[0], y=location[1], canvas=ax.figure.canvas)
ax.figure.canvas.draw()
cursor.onmove(event)
fig.savefig('text_cursor.png')
```

and added the note. Sending the fake event AND having plt.show also does not work. So I removed all that code and did not tried to add e.g. a "draw" event callback, because that might confuse the reader and one might have trouble distinguishing between stuff one needs for the core example and stuff needed for the sphinx gallery.

So I came to the conclusion: It's not worth it.

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+"""
+==========
+Textcursor
```
```suggestion
Annotated cursor
```
or similar.

Textcursor is often associtated with a cursor for text selection:
![grafik](https://user-images.githubusercontent.com/2836374/115466120-e8555080-a22f-11eb-9eec-98da271a81e6.png)

Joschua-Conrad replied:
Renamed .py file and the class itself accordingly.

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+            super()._update()
+
+
+fig = plt.figure(figsize=(8, 6))
+ax = fig.add_subplot(111, facecolor='#FFFFCC')
```
```suggestion
fig, ax = plt.subplots(figsize=(8, 6))
```
Is a canonical way. It's good to keep examples minimal to not clutter the relevant code.
Also I try to move away from that 90ish background color in the widget examples.

Sorry that the base Cursor example is not a good template here.

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+fig = plt.figure(figsize=(8, 6))
+ax = fig.add_subplot(111, facecolor='#FFFFCC')
+
+# A linearly growing x vector.
+x = np.linspace(-5, 5, 1000)
+
+# A non biunique function. dataaxis='y' will cause trouble.
+y = (x ** 2)
```
```suggestion
x = np.linspace(-5, 5, 1000)
y = x**2
```
is sufficient. No need to explain the obvious ("linearly growing x vector").

If you really want, add a `# sample data` comment on top.

I don't understand "dataaxis='y' will cause trouble.", but seems out of context here.

Joschua-Conrad replied:
You're right, the comment is indeed out of context.

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+# A non biunique function. dataaxis='y' will cause trouble.
+y = (x ** 2)
+
+lin = ax.plot(x, y)
```
```suggestion
line, = ax.plot(x, y)
```
Tuple unpacking as a canonical way. No need to save letters; we have plenty of them 😄 .

Joschua-Conrad replied:
Suggested change applied. Had to change renamed lin => line also at another spot, so I could not just commit the exact suggestion you posted.

timhoffm commented on examples/widgets/text_cursor.py:
```diff
@@ -0,0 +1,313 @@
+    color='red',
+    linewidth=2)
+
+# A call demonstrating problems with the dataaxis=y parameter.
+# The text now looks up the matching x value for the current cursor y position
+# instead of vice versa. Hover you cursor to y=4. There are two x values
+# producing this y value: -2 and 2. The function is only unique,
+# but not biunique. Only one value is shown in the text.
+# cursor = TextCursor(
+#     line=lin[0],
+#     numberformat="{0:.2f}\n{1:.2f}",
+#     dataaxis='y', offset=[10, 10],
+#     textprops={'color': 'blue', 'fontweight': 'bold'},
+#     ax=ax,
+#     useblit=True,
+#     color='red', linewidth=2)
```
Not sure if you want to go through this discussion and the added complexity within the example. Alternatively, you could assume that x is the independent variable as is the case for most plots, and let users figure out the other case.

If this stays, this should be moved to a separate paragraph below `plt.show()`. See https://sphinx-gallery.github.io/stable/syntax.html#how-to-structure-your-python-scripts-for-sphinx-gallery

Joschua-Conrad replied:
I now have this code organized in a second plt.show() and like it's described in the sphinx gallery docs you linked. Can this stay the way I organized it now, or shall I remove the whole hint?

timhoffm reviewed (commented):
See also https://matplotlib.org/stable/gallery/misc/cursor_demo.html#sphx-glr-gallery-misc-cursor-demo-py, which is related. It implements a cursor without relying on the cursor widget, but otherwise has some similar features like your example. Not sure how to organize all this reasonably.  Suggestions are welcome.

Joschua-Conrad commented:
> See also https://matplotlib.org/stable/gallery/misc/cursor_demo.html#sphx-glr-gallery-misc-cursor-demo-py, which is related. It implements a cursor without relying on the cursor widget, but otherwise has some similar features like your example. Not sure how to organize all this reasonably. Suggestions are welcome.

I now reference the cross hair cursor in my example. But I'd say that the annotated cursor is different, because it inherits from a default widget and therefore shows how to overwrite the callbacks without crashing blit() by restoring the background after drawing each sub-element (cursor lines and text).

The cursor @timhoffm linked skips redrawing the cursor when the mouse moves between two plotted points. I added this feature in ebc3f98c7f4962057626cace05d3ff08e2bb1180.

@timhoffm Thank you for your detailed review! I addressed all remarks and either closed the remark, when I just did exactly what you mentioned and left the discussion open when I changed something, but am not sure if my change resolves the remark.

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        # Get the coordinates, which should be displayed as text,
+        # if the event cooridnates are valid
+        plotpoint = None
+        if (event.xdata is not None) and (event.ydata is not None):
```
```suggestion
        if event.xdata is not None and event.ydata is not None:
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        # Skip even the call of the base class, because this would restore the
+        # background, draw the cursor lines and would leave us the job to
+        # re-draw the text.
+        if plotpoint is not None:
+            if plotpoint == self.lastdrawnplotpoint:
+                return
```
```suggestion
        if plotpoint is not None and plotpoint == self.lastdrawnplotpoint:
            return
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        # Check if the display of text is still necessary.
+        # If not, just return.
+        # This behaviour is also cloned from the base class.
+        if (not self.get_active()) or (not self.visible):
```
```suggestion
        if not self.get_active() or not self.visible:
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+            data = ydata
+            lim = self.ax.get_ylim()
+        else:
+            raise ValueError(
+                "The data axis specifier {} should be 'x' or 'y'"
+                .format(self.dataaxis))
```
```suggestion
            raise ValueError(f"The data axis specifier {self.dataaxis} should "
                             f"be 'x' or 'y'")
```

Joschua-Conrad replied:
I never knew that string format syntax, but I love it :D

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+                "The data axis specifier {} should be 'x' or 'y'"
+                .format(self.dataaxis))
+
+        # If position is valid
+        if pos is not None:
+            # And in valid plot data range
+            if pos >= lim[0] and pos <= lim[-1]:
+                # Convert given positon to numpy array,
+                # so numpy function can be used.
+                findme = np.array([pos])
+                # Find closest x value in sorted x vector.
+                # This is the code line,
+                # which requires the plotted data to be sorted.
+                index = np.searchsorted(data, findme)
+                # Return none, if this index is out of range.
+                if (index < 0) or (index >= len(data)):
+                    return None
+                # Return plot point as tuple.
+                return (xdata[index][0], ydata[index][0])
```
```suggestion
        # If position is valid and in valid plot data range.
        if pos is not None and lim[0] <= pos <= lim[-1]:
            # Find closest x value in sorted x vector.
            # This requires the plotted data to be sorted.
            index = np.searchsorted(data, pos)
            # Return none, if this index is out of range.
            if index < 0 or index >= len(data):
                return None
            # Return plot point as tuple.
            return (xdata[index][0], ydata[index][0])
```

Joschua-Conrad replied:
Commit d027645ba81649eccaa127a739235cae5efc8c0a fixes a bug introduced with this change: Had to remove [0] in the returned tuple

QuLogic replied:
Right, I forgot to remove those.

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        return None
+
+    def clear(self, event):
+        """Overwirtten clear callback for cursor. Called right before \
+        displaying the figure."""
```
```suggestion
        """
        Overridden clear callback for cursor, called before drawing the figure.
        """
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        self.lastdrawnplotpoint = None
+
+    def onmove(self, event):
+        """Overwritten draw callback for cursor. \
+        Called when moving the mouse."""
```
```suggestion
        """
        Overridden draw callback for cursor, called when moving the mouse.
        """
```

Joschua-Conrad replied:
Fixed in multiple spots in 4aef9a8dbd0e5b4489255c85e2d0a8f7d62b4c25

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        """Overwritten draw callback for cursor. \
+        Called when moving the mouse."""
+
+        # Leave method under the same conditions as in overwritten method
```
```suggestion
        # Leave method under the same conditions as in overridden method.
```

Joschua-Conrad replied:
Fixed in multiple spots in 4aef9a8dbd0e5b4489255c85e2d0a8f7d62b4c25

Joschua-Conrad replied:
Linebreak added in 4d3dde10acc80afc865315b6c14416c47049aaa7

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+            return
+
+        # Get the coordinates, which should be displayed as text,
+        # if the event cooridnates are valid
```
```suggestion
        # if the event coordinates are valid.
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+
+        # Baseclass redraws canvas and cursor. Due to blitting,
+        # the added text is removed in this call, because the
+        # backround is redrawn.
```
```suggestion
        # background is redrawn.
```

Joschua-Conrad replied:
Fixed in 0a740fc8cad97a88207777af38618adcac8bd744 in multiple spots

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+        if (not self.get_active()) or (not self.visible):
+            return
+
+        # Draw the widget, if event coordinates are valid
+        # If plotpoint is valid (not None)
```
```suggestion
        # Draw the widget, if event coordinates are valid.
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+            self.text.set_visible(False)
+
+        # Draw changes. Canot use _update method of baseclass,
+        # because it would first restore the backround, which
```
```suggestion
        # because it would first restore the background, which
```

Joschua-Conrad replied:
Fixed in 0a740fc8cad97a88207777af38618adcac8bd744 in multiple spots

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+            self.ax.draw_artist(self.text)
+            self.canvas.blit(self.ax.bbox)
+        else:
+            # If blitting is deactivated, the overwritten _update call made
```
```suggestion
            # If blitting is deactivated, the overridden _update call made
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+            # We still have to draw the changes.
+            self.canvas.draw_idle()
+
+    def setpos(self, xpos, ypos):
```
```suggestion
    def set_position(self, xpos, ypos):
```

Joschua-Conrad replied:
Fixed in fee740e298d6a9bb863dd452900c43b8cc97c98f in multiple spots

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+
+        Returns
+        -------
+        ret : {2D array-like, None} The coordinates which should be displayed.
+            *None* is the fallback value.
```
```suggestion
        ret : {2D array-like, None}
            The coordinates which should be displayed.
            *None* is the fallback value.
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+
+    def _update(self):
+        """
+        Overwritten method for eather blitting or drawing the widget canvas.
```
```suggestion
        Overridden method for either blitting or drawing the widget canvas.
```

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,345 @@
+# cursor = AnnotatedCursor(line=lin[0], ax=ax, useblit=True)
+
+# A more advanced call. Properties for text and lines are passed.
+# See the color if you are confused which parameter is passed where.
```
What do you mean "See the color"?

Joschua-Conrad replied:
Imporved the explanation in b1c0b2498351827394c1229c18e7f28e0bcc5ea4

QuLogic reviewed (commented):
Please squash this when merging.

Joschua-Conrad commented:
> Please squash this when merging.

Will do.

@QuLogic all your requested changes were clear and applied. Maybe check in b1c0b2498351827394c1229c18e7f28e0bcc5ea4, if the new explanation is coherent.

Sorry for launching an avalanche of workflows, but I don't know how to suppress workflows for the first few commits when many commits are created within a short time...

jklymak commented:
you can't suppress the workflows unfortunately, but you have 45 commits.  Can you squash these?  If that is a pain we will squash them anyway before merging, so there is really not a lot of need to keep so many commits.

QuLogic commented:
> Sorry for launching an avalanche of workflows, but I don't know how to suppress workflows for the first few commits when many commits are created within a short time...

You can batch commit suggested changes from the diff view, or just do the work locally.

Joschua-Conrad commented:
> you can't suppress the workflows unfortunately, but you have 45 commits. Can you squash these? If that is a pain we will squash them anyway before merging, so there is really not a lot of need to keep so many commits.

I referenced some of the commits in my replies and I can only squash the commits be squash-merging them to a new branch, but I then cannot change the branch to which this PR is attached. So I'd keep the commits and squash them when merging into master.

Joschua-Conrad commented:
> > Sorry for launching an avalanche of workflows, but I don't know how to suppress workflows for the first few commits when many commits are created within a short time...
> 
> You can batch commit suggested changes from the diff view, or just do the work locally.

@QuLogic Thank you, the change-approval in the diff view is what I was looking for!

Joschua-Conrad commented:
Do I still need to change or answer something? I answered questions, applied suggested changes and left two answers, where I'm not sure open, which are:

- https://github.com/matplotlib/matplotlib/pull/20019#discussion_r617350188
- https://github.com/matplotlib/matplotlib/pull/20019#discussion_r617352012

Is there something else I can do?

Joschua-Conrad commented:
Do I still need to change or answer something? I answered questions, applied suggested changes and left two answers, where I'm not sure open, which are:

- https://github.com/matplotlib/matplotlib/pull/20019#discussion_r617350188
- https://github.com/matplotlib/matplotlib/pull/20019#discussion_r617352012

Is there something else I can do?

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,340 @@
+# -----------------------------------
+# A call demonstrating problems with the *dataaxis=y* parameter.
+# The text now looks up the matching x value for the current cursor y position
+# instead of vice versa. Hover you cursor to y=4. There are two x values
```
```suggestion
# instead of vice versa. Hover your cursor to y=4. There are two x values
```

Joschua-Conrad commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,340 @@
+        super().__init__(**cursorargs)
+
+        # The numbers displayed next to cursor. Default value.
+        self.textdata = [0, 0]
```
Sorry for forgetting to remove this and thank you for noticing!
Fixed in c2ba60b69e3ca7538ab75ae43219b937776e1741

QuLogic commented:
You will need to fix the flake8 issues before this is merged.

Joschua-Conrad commented:
> 
> 
> You will need to fix the flake8 issues before this is merged.

Sorry I had my editor set up for another project, where line length issues are ignored during linting.

Joschua-Conrad commented:
> You will need to fix the flake8 issues before this is merged.

@QuLogic All checks pass now. Thank you for the hint on that!

QuLogic commented on examples/widgets/annotated_cursor.py:
```diff
@@ -0,0 +1,338 @@
+# producing this y value: -2 and 2. The function is only unique,
+# but not biunique. Only one value is shown in the text.
+
+fig, ax = plt.subplots(figsize=(8, 6))
```
I would add a (short) title on this Axes, because if you just run the example without looking at it, you might wonder why it's not working. Maybe also on the first figure as well.

Joschua-Conrad replied:
Done in 68a16913747277b417991cce2e5efd856d36c7a3
I added a ax.set_title("Cursor Tracking [x|y] Position") call for both figures.

QuLogic commented:
Thanks @Joschua-Conrad ! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

Joschua-Conrad commented:
Thank you all for helping me with making my code mergable!
